src/utils/helpers/endpoints/endpointUrlBuilder.ts

```diff
@@ -0,0 +1,144 @@
+import { BasicObject } from '@openmsupply/expression-evaluator/lib/types'
+import config from '../../../config'
+import { RestEndpoints } from './types'
+
+const {
+  isProductionBuild,
+  restEndpoints,
+  localHostREST,
+  localHostGraphQL,
+  productionPathREST,
+  productionPathGraphQL,
+} = config
+const { port, hostname, protocol } = window.location
+const getProductionUrl = (path: string) => `${protocol}//${hostname}:${port}${path}`
+
+const serverREST = isProductionBuild ? getProductionUrl(productionPathREST) : localHostREST
+export const serverGraphQL = isProductionBuild
+  ? getProductionUrl(productionPathGraphQL)
+  : localHostGraphQL
+
+const getServerUrl = (...args: RestEndpoints | ['graphQL']): string => {
+  // "as" here ensures we must have types/cases for ALL keys of
+  // config.restEndpoints
+  const endpointKey = args[0] as keyof typeof restEndpoints | 'graphQL'
+  if (endpointKey === 'graphQL') return serverGraphQL
+  const endpointPath = restEndpoints[endpointKey]
+
+  switch (endpointKey) {
+    case 'public':
+    case 'prefs':
+    case 'login':
+    case 'loginOrg':
+    case 'userInfo':
+    case 'createHash':
+    case 'generatePDF':
+    case 'admin':
+    case 'installLanguage':
+    case 'allLanguages':
+    case 'previewActions':
+    case 'extendApplication':
+      return serverREST + endpointPath
+
+    case 'language':
+      const languageCode = args[1]
+      return `${serverREST}${endpointPath}/${languageCode}`
+
+    case 'file':
+      const fileId = args[1]
+      const isThumbnail = args[2] === 'thumbnail'
+      return `${serverREST}${endpointPath}?uid=${fileId}${isThumbnail ? '&thumbnail=true' : ''}`
+
+    case 'verify':
+      const uid = args[1]
+      return `${serverREST}${endpointPath}?uid=${uid}`
+
+    case 'userPermissions':
+      const permissionParameters = args[1]
+      return `${serverREST}${endpointPath}${buildQueryString(permissionParameters as BasicObject)}`
+
+    case 'checkTrigger':
+      const serial = args[1]
+      return `${serverREST}${endpointPath}?serial=${serial}`
+
+    case 'checkUnique':
+      const checkUniqueParams = args[1]
+      return `${serverREST}${endpointPath}${buildQueryString(checkUniqueParams as BasicObject)}`
+
+    case 'dataViews':
+      const tableName = args[1]
+      const itemId = typeof args[2] === 'string' ? args[2] : undefined
+      const query = typeof args[2] === 'object' ? args[2] : undefined
+      // Detail view
+      if (itemId) return `${serverREST}${endpointPath}/table/${tableName}/item/${itemId}`
+      // Table view
+      if (tableName)
```

Comment:
I think this should be at the top
```
if (!tableName)
return `${serverREST}${endpointPath}`
```

Reply:
Cool. This plus above suggestions: b5887213